art/utils_test.py

```diff
@@ -210,38 +211,109 @@ def get_classifier_kr(loss_name='categorical_crossentropy'):
-        loss = keras.losses.categorical_hinge
+        if loss_type == 'label':
+            raise AttributeError
+        elif loss_type == 'function_losses':
+            loss = keras.losses.categorical_hinge
     elif loss_name == 'categorical_crossentropy':
-        loss = keras.losses.categorical_crossentropy
+        if loss_type == 'label':
+            if from_logits:
+                raise AttributeError
+            else:
+                loss = loss_name
+        elif loss_type == 'function_losses':
+            if from_logits:
+                if int(keras.__version__.split('.')[0]) == 2 and int(keras.__version__.split('.')[1]) >= 3:
+                    def categorical_crossentropy(y_true, y_pred):
+                        return keras.losses.categorical_crossentropy(y_true, y_pred, from_logits=True)
+
+                    loss = categorical_crossentropy
+                else:
+                    raise AttributeError
+            else:
+                loss = keras.losses.categorical_crossentropy
+        elif loss_type == 'function_backend':
+            if from_logits:
+                def categorical_crossentropy(y_true, y_pred):
+                    return keras.backend.categorical_crossentropy(y_true, y_pred, from_logits=True)
+
+                loss = categorical_crossentropy
+            else:
+                loss = keras.backend.categorical_crossentropy
     elif loss_name == 'sparse_categorical_crossentropy':
-        loss = keras.losses.sparse_categorical_crossentropy
-    elif loss_name == 'binary_crossentropy':
-        loss = keras.losses.binary_crossentropy
+        if loss_type == 'label':
+            if from_logits:
+                raise AttributeError
+            else:
+                loss = loss_name
+        elif loss_type == 'function_losses':
+            if from_logits:
+                if int(keras.__version__.split('.')[0]) == 2 and int(keras.__version__.split('.')[1]) >= 3:
+                    def sparse_categorical_crossentropy(y_true, y_pred):
+                        return keras.losses.sparse_categorical_crossentropy(y_true, y_pred, from_logits=True)
+
+                    loss = sparse_categorical_crossentropy
+                else:
+                    raise AttributeError
+            else:
+                loss = keras.losses.sparse_categorical_crossentropy
+        elif loss_type == 'function_backend':
+            if from_logits:
+                def sparse_categorical_crossentropy(y_true, y_pred):
+                    return keras.backend.sparse_categorical_crossentropy(y_true, y_pred, from_logits=True)
+
+                loss = sparse_categorical_crossentropy
+            else:
+                loss = keras.backend.sparse_categorical_crossentropy
     elif loss_name == 'kullback_leibler_divergence':
-        loss = keras.losses.kullback_leibler_divergence
-    elif loss_name == 'cosine_proximity':
-        loss = keras.losses.cosine_proximity
+        if loss_type == 'label':
```

Comment:
Some of the if-elses around `loss_type` can probably be avoided by introducing a `dict` that maps the loss type to the action, then doing something like `loss = dict_loss_type[loss_type]`. The dict value could even be a function that does some of the checks required. This goes for all constructs of this type in the test models.

Reply:
I agree and have tried to create a solution based on a dictionary but I feels like it becomes a quite complicated dictionary. I tried to improve the two functions by extending the docstrings and adding exception messages.